p2p/transport/webrtc/stream.go

```diff
@@ -0,0 +1,235 @@
+package libp2pwebrtc
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+	"github.com/libp2p/go-msgio/pbio"
+	"github.com/pion/datachannel"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ network.MuxedStream = &webRTCStream{}
+
+const (
+	// maxMessageSize is limited to 16384 bytes in the SDP.
+	maxMessageSize int = 16384
+	// Pion SCTP association has an internal receive buffer of 1MB (roughly, 1MB per connection).
+	// We can change this value in the SettingEngine before creating the peerconnection.
+	// https://github.com/pion/webrtc/blob/v3.1.49/sctptransport.go#L341
+	maxBufferedAmount int = 2 * maxMessageSize
+	// bufferedAmountLowThreshold and maxBufferedAmount are bound
+	// to a stream but congestion control is done on the whole
+	// SCTP association. This means that a single stream can monopolize
+	// the complete congestion control window (cwnd) if it does not
+	// read stream data and it's remote continues to send. We can
+	// add messages to the send buffer once there is space for 1 full
+	// sized message.
+	bufferedAmountLowThreshold uint64 = uint64(maxBufferedAmount) / 2
+
+	// Proto overhead assumption is 5 bytes
+	protoOverhead int = 5
+	// Varint overhead is assumed to be 2 bytes. This is safe since
+	// 1. This is only used and when writing message, and
+	// 2. We only send messages in chunks of `maxMessageSize - varintOverhead`
+	// which includes the data and the protobuf header. Since `maxMessageSize`
+	// is less than or equal to 2 ^ 14, the varint will not be more than
+	// 2 bytes in length.
+	varintOverhead int = 2
+)
+
+// Package pion detached data channel into a net.Conn
+// and then a network.MuxedStream
+type (
+	webRTCStream struct {
+		reader webRTCStreamReader
+		writer webRTCStreamWriter
+
+		stateHandler webRTCStreamState
+
+		// hack for closing the Read side using a deadline
+		// in case `Read` does not return.
+		closeErr    error
+		closeErrMux sync.Mutex
+
+		conn *connection
+		id   uint16
+		rwc  datachannel.ReadWriteCloser
+
+		laddr net.Addr
+		raddr net.Addr
+
+		wg sync.WaitGroup
+
+		ctx    context.Context
+		cancel context.CancelFunc
+
+		closeOnce sync.Once
+	}
+)
+
+func newStream(
+	connection *connection,
+	channel *webrtc.DataChannel,
+	rwc datachannel.ReadWriteCloser,
+	laddr, raddr net.Addr,
+) *webRTCStream {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	reader := bufio.NewReaderSize(rwc, maxMessageSize)
```

Comment:
Will add a comment. Reason being that we also write up to 16KiB of data. And Pion does not allow data to be read in chunks, e.g. if you read less data then there was send, it will just drop the other part.

Reply:
Going to check with @ckousik .

Reply:
I've removed it for now as I couldn't find something myself and all tests work as fine without this code as before.

Reply:
The writing part does not matter here, but our max message size defined by the SDP is 16kB. Pion does not allow reading a partial message, i.e. if we read 4k from an 8k message, it will drop the remaining part and fail with `ErrShortBuffer`. Since the largest message we could receive is 16kB, we need to have the allocation.